src/cata_imgui.cpp

```diff
@@ -0,0 +1,466 @@
+        text_align alignment, float max_width, bool *is_selected, bool *is_focused, bool *is_hovered )
+{
+    ImGui::PushID( text.c_str() );
+    ImGuiID itemId = GImGui->CurrentWindow->IDStack.back();
+    const auto color_segments = split_by_color( text );
+    std::stack<nc_color> color_stack;
+    color_stack.push( color );
+    size_t chars_per_line = size_t( max_width );
+    if( chars_per_line == 0 ) {
+        chars_per_line = SIZE_MAX;
+    }
+    float cursor_start_x = ImGui::GetCursorPosX();
+#if defined(WIN32) || defined(TILES)
+    size_t char_width = size_t( ImGui::CalcTextSize( " " ).x );
+    chars_per_line /= char_width;
+#endif
+    if( alignment != text_align::Left ) {
+        std::string str_raw = remove_color_tags( text );
+        int fullWidth = ImGui::GetContentRegionAvail().x;
+        float textWidth = ImGui::CalcTextSize( str_raw.c_str() ).x;
+        if( alignment == text_align::Right ) {
+            ImGui::SetCursorPosX( ImGui::GetCursorPosX() + fullWidth - textWidth - 2 );
+        } else if( alignment == text_align::Center ) {
+            ImGui::SetCursorPosX( ImGui::GetCursorPosX() + ( float( fullWidth ) / 2 ) - ( textWidth / 2 ) );
+        }
+    }
+    if( is_selected != nullptr ) {
+        ImGui::Selectable( "", is_selected );
+        ImGui::SameLine( 0, 0 );
+    }
+
+    int i = 0;
+    size_t current_x = 0;
+    for( auto seg : color_segments ) {
+        if( seg.empty() ) {
+            continue;
+        }
+
+        if( seg[0] == '<' ) {
+            const color_tag_parse_result::tag_type type =
+                update_color_stack( color_stack, seg, report_color_error::yes );
+            if( type != color_tag_parse_result::non_color_tag ) {
+                seg = rm_prefix( seg );
+            }
+        }
+
+        color = color_stack.empty() ? color : color_stack.top();
+        for( size_t current_seg_index = 0; current_seg_index < seg.length(); ) {
+
+            if( i++ > 0 ) {
+                if( current_x != 0 ) {
+                    ImGui::SameLine( 0, 0 );
+                } else if( alignment == text_align::Left ) {
+                    ImGui::SetCursorPosX( cursor_start_x );
+                }
+            }
+            size_t chars_to_print = seg.length() - current_seg_index;
+            if( alignment == text_align::Left ) {
+                chars_to_print = std::min( chars_per_line - current_x, chars_to_print );
+            }
+#if !(defined(TILES) || defined(WIN32))
+            int pair_id = color.get_index();
+            pairs &pair = colorpairs[pair_id];
+
+            int palette_index = pair.FG != 0 ? pair.FG : pair.BG;
+            if( color.is_bold() ) {
+                palette_index += color_loader<RGBTuple>::COLOR_NAMES_COUNT / 2;
+            }
+            RGBTuple &rgbCol = rgbPalette[palette_index];
+            ImGui::TextColored( { static_cast<float>( rgbCol.Red / 255. ), static_cast<float>( rgbCol.Green / 255. ),
```

Comment:
How does ImGUI handle rgb values not supported by a terminal?

Reply:
ImTui converts the rgb values to ncurses friendly values by some method, if you don't get the exact color you get the closest approximation, if the color pair isn't currently uploaded to ncurses then it uploads it.